### PR TITLE
8304976: Optimize DateTimeFormatterBuilder.ZoneTextPrinterParser.getTree()

### PR DESCRIPTION
**Pull request by stsypanov**
1) When `DateTimeFormatter` is reused we don't need to copy `availableZoneIds` and allocate `nonRegionIds` as PrefixTree can be taken from cache. In the related benchmark allocation of `HashSet` takes ~93% of all time, so avoiding it should bring some improvement for cases when we reuse `DateTimeFormatter` to parse a string into `ZoneDateTime`
image

2) `DateTimeFormatter` is mostly used with one locale, so `cachedTree` and `cachedTreeCI` can have predefined size.

```java
@State(Scope.Thread)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class DateTimeFormatterBenchmark {

    private static final DateTimeFormatter df = new DateTimeFormatterBuilder().appendPattern("yyyy:MM:dd:HH:mm:v").toFormatter();
    private static final String TEXT = "2015:03:10:12:13:ECT";

    @Setup
    public void setUp() {
        ZonedDateTime.parse(TEXT, df);
    }

    @Benchmark
    public ZonedDateTime parse() {
        return ZonedDateTime.parse(TEXT, df);
    }
}
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304976](https://bugs.openjdk.org/browse/JDK-8304976): Optimize DateTimeFormatterBuilder.ZoneTextPrinterParser.getTree()

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12612/head:pull/12612` \
`$ git checkout pull/12612`

Update a local copy of the PR: \
`$ git checkout pull/12612` \
`$ git pull https://git.openjdk.org/jdk.git pull/12612/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12612`

View PR using the GUI difftool: \
`$ git pr show -t 12612`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12612.diff">https://git.openjdk.org/jdk/pull/12612.diff</a>

</details>

### CODE REVIEW

**Comment by stsypanov** [comment]
Benchmarking
```
make test TEST="micro:time.format.ZonedDateTimeFormatterBenchmark" MICRO="OPTIONS=-prof gc"

master

Benchmark                                                              Mode  Cnt      Score     Error   Units
ZonedDateTimeFormatterBenchmark.parse                                  avgt   20  13545.737 ± 451.164   ns/op
ZonedDateTimeFormatterBenchmark.parse:·gc.alloc.rate                   avgt   20   1713.579 ±  56.517  MB/sec
ZonedDateTimeFormatterBenchmark.parse:·gc.alloc.rate.norm              avgt   20  24320.006 ±   0.001    B/op
ZonedDateTimeFormatterBenchmark.parse:·gc.count                        avgt   20    110.000            counts
ZonedDateTimeFormatterBenchmark.parse:·gc.time                         avgt   20    514.000                ms


patch

Benchmark                                                              Mode  Cnt     Score    Error   Units
ZonedDateTimeFormatterBenchmark.parse                                  avgt   20  1276.063 ± 52.543   ns/op
ZonedDateTimeFormatterBenchmark.parse:·gc.alloc.rate                   avgt   20   622.223 ± 26.456  MB/sec
ZonedDateTimeFormatterBenchmark.parse:·gc.alloc.rate.norm              avgt   20   832.001 ±  0.001    B/op
ZonedDateTimeFormatterBenchmark.parse:·gc.count                        avgt   20    58.000           counts
ZonedDateTimeFormatterBenchmark.parse:·gc.time                         avgt   20   475.000               ms
```

File: test/micro/org/openjdk/bench/java/time/format/ZonedDateTimeFormatterBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package time.format;
```
Comment (naotoj) [inline test/micro/org/openjdk/bench/java/time/format/ZonedDateTimeFormatterBenchmark.java:23]: package should be `org.openjdk.bench.java.time.format`?

**Review by naotoj — CHANGES_REQUESTED** [review]
Looks like a nice fix to me.

**Review by naotoj — APPROVED** [review]
LGTM

**Pull request closed without merging**